src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1927 @@
+                if (token.IsMissing ||
+                    token.Diagnostics.Length > 0)
+                {
+                    return true;
+                }
+
+                foreach (var trivia in token.LeadingTrivia)
+                {
+                    if (trivia.Diagnostics.Length > 0)
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private RegexExpressionNode ParseRightSideOfCharacterClassRange()
+        {
+            // Parsing the right hand side of a - is extremely strange (and most likely  buggy) in 
+            // the .net parser. Specifically, the .net parser will still consider itself on the right
+            // side no matter how many escaped dashes it sees.  So, for example, the following is legal
+            // [a-\-] (even though \- is less than 'a'). Similarly, the following are *illegal* 
+            // [b-\-a] and [b-\-\-a].  That's because the range that is checked is actually "b-a", even
+            // though it has all the \- escapes in the middle.
+
+            var first = ParseSingleCharacterClassComponent(isFirst: false, afterRangeMinus: true);
+            if (!IsEscapedMinus(first))
+            {
+                return first;
+            }
+
+            var builder = ArrayBuilder<RegexExpressionNode>.GetInstance();
+            builder.Add(first);
+
+            while (IsEscapedMinus(builder.Last()) && _currentToken.Kind != RegexKind.CloseBracketToken)
+            {
+                builder.Add(ParseSingleCharacterClassComponent(isFirst: false, afterRangeMinus: true));
+            }
+
+            return new RegexSequenceNode(builder.ToImmutable());
+        }
+
+        private RegexPrimaryExpressionNode ParseSingleCharacterClassComponent(bool isFirst, bool afterRangeMinus)
+        {
+            if (_currentToken.Kind == RegexKind.BackslashToken && _lexer.Position < _lexer.Text.Length)
+            {
+                var backslashToken = _currentToken;
+                var afterSlash = _lexer.Position;
+
+                // trivia is not allowed anywhere in a character class, and definitely not between
+                // a \ and the following character.
+                ConsumeCurrentToken(allowTrivia: false);
+                Debug.Assert(_currentToken.VirtualChars.Length == 1);
+
+                var nextChar = _currentToken.VirtualChars[0].Char;
+                switch (nextChar)
+                {
+                    case 'D': case 'd':
+                    case 'S': case 's':
+                    case 'W': case 'w':
+                    case 'p': case 'P':
+                        if (afterRangeMinus)
+                        {
+                            backslashToken = backslashToken.AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                                string.Format(WorkspacesResources.Cannot_include_class_0_in_character_range, nextChar),
```

Comment:
The logic and message don't match. Should the message say "cannot include class {0] on the right of a character range"?

Reply:
Correct.  This is a case where we're matching the message of hte native regex engine.